### PR TITLE
allow default system namespace to be overridden using an env var

### PR DESCRIPTION
**Pull request by rawlingsj**
I'd like to install Build Pipeline in namespace other than `tekton-pipelines`.  This change still uses `tekton-pipelines` as the default but allows folks to set an environment variable to override.  As an example I've hacked a Helm chart for easy install while integrating with Build Pipeline, this uses the downward api to set the new env var to the current namespace.

If it's of interest [this is the chart](https://github.com/jenkins-x-charts/knative-build-pipeline/tree/master/knative-build-pipeline) and override the new env var [here](https://github.com/jenkins-x-charts/knative-build-pipeline/blob/master/knative-build-pipeline/templates/controller.yaml#L41-L44).

Also deleted unused `pkg/names.go` as I think it was duplicated in `pkg/system/names.go`.

### CODE REVIEW

**Comment by rawlingsj** [comment]
I probably ought to add the new env var to the docs somewhere too, in https://github.com/knative/build-pipeline/blob/master/DEVELOPMENT.md#environment-setup maybe?

**Comment by bobcatfish** [comment]
Thanks for this @rawlingsj !! 🎉 

> I probably ought to add the new env var to the docs somewhere too

Definitely agree! I think what we really need is some docs about how to setup and configure pipelines, which we don't have yet 🤔 (kind of related to #385) 

So some options:
* pave new ground + start a new doc on setting up pipelines (with almost nothing in it for now - you could link to your sweet helm chart tho!!)
* Put it in the DEVELOPMENT.md for now like you suggested
* Add it to [these (sparse) docs we have on configuring the entrypoint image](https://github.com/knative/build-pipeline/blob/master/docs/using.md#configure-entrypoint-image)

> but allows folks to set an environment variable to override

What do you think about configuring the namespace in a config-map instead of an environment variable? Just wondering cuz we already have a config-map for some other controller config. Maybe this doesn't make sense tho b/c the controller wouldn't even know what namespace to look for the namespace in 🤔

**Comment by bobcatfish** [comment]
@shashwathi and @ImJasonH pointed out that there have been similar changes to serving: https://github.com/knative/serving/pull/2708 <-- so as long as we do pretty much the same thing as that, we should be good! and it looks I think like this configuration also ends up using an env var

**Comment by markusthoemmes** [comment]
FWIW, a generalized version of the serving bit was added to knative/pkg recently. I think all projects should be able to use that? https://github.com/knative/pkg/pull/235

**Comment by rawlingsj** [comment]
ok great - thanks @bobcatfish @markusthoemmes - will take a look

**Comment by rawlingsj** [comment]
sorry for the delay but juggling lots of things ATM, so we're suggesting we 
- add a dependency on `knative/pkg` repo
- use https://github.com/knative/pkg/blob/8e50d82/system/names.go#L30 instead of the same approach in this PR
- [panic](https://github.com/knative/pkg/blob/8e50d82/system/names.go#L35) if the env var isn't set using the downward api, which is different from existing behaviour where a default `namespace` is used
- switch the `./config/*.yaml` files in this repo to use the downward api and remove the `namespace: knative-build-pipeline` from all yamls, example https://github.com/knative/build-pipeline/blob/01853c6/config/200-serviceaccount.yaml#L18
- delete the create [namespace yaml](https://github.com/knative/build-pipeline/blob/01853c6/config/100-namespace.yaml) and add to install docs to create a new namespace with `kubectl` if folks don't want it in the current namespace?

If that's all correct and folks are happy I can close this PR and create a new with above?

**Comment by rawlingsj** [comment]
Hmm - just started and suspect there's going to be a problem updating the `./config/*.yaml` because there's a `ClusterRoleBinding` that needs the namespace that the serviceaccount lives in https://github.com/knative/build-pipeline/blob/01853c6/config/201-clusterrolebinding.yaml#L21, we get around this in helm world by templating the namespace https://github.com/jenkins-x-charts/knative-build-pipeline/blob/b8d1473/knative-build-pipeline/templates/200-clusterrolebinding.yaml#L21 so it's set at install time to the current namespace just as the downward api is set at install.

I'm not sure how we can use https://github.com/knative/pkg/blob/8e50d82/system/names.go#L30 from this repo and keep the current approach of installing via `ko apply -f config/`,  I think there needs to be some kinda templating on the `ClusterRoleBinding` to set the current namespace or update docs to change it which doesn't seem great.

Any other thoughts or ideas?

**Comment by rawlingsj** [comment]
Might be nice to avoid the `ClusterRoleBinding`s all together and create `RoleBindings` instead as that would get around the problem, not sure if the knative build pipeline need to be cluster wide role bindings?

**Comment by abayer** [comment]
Good question - @bobcatfish, do you know re ^^^?

**Comment by bobcatfish** [comment]
> I'm not sure how we can use https://github.com/knative/pkg/blob/8e50d82/system/names.go#L30 from this repo and keep the current approach of installing via `ko apply -f config/`, I think there needs to be some kinda templating on the `ClusterRoleBinding` to set the current namespace or update docs to change it which doesn't seem great.

Yeah this is weird, im looking at knative/serving to see how they've done it, since they seem to be using `system.Namespace()` (e.g. [here](https://github.com/knative/serving/blob/master/pkg/reconciler/v1alpha1/revision/resources/queue.go#L141)), but they've still go the namespace [hardcoded in all their yaml files](https://github.com/knative/serving/tree/master/config) (and [the config to create the namespace still exists](https://github.com/knative/serving/blob/master/config/100-namespace.yaml))

Looking at the PR where this was added https://github.com/knative/serving/pull/2708 I get the impression there was no good story at that point either about how to handle those yamls, it seems like the assumption is that a user will change them manually :thinking: 

My only recommendation is to leave the namespace hard-coded in our .yaml files for now, and add some docs about how folks can change them manually (and use the downward api) - not sure there are any better options (maybe @ImJasonH has some ideas)

**Comment by ImJasonH** [comment]
@bobcatfish's suggestion about leaving hard-coded values in YAML and documenting changing it SGTM, especially if that's what Serving has today. When in doubt we should go along with what the other repos do to solve these problems, even if those solutions are half-baked.

**Comment by rawlingsj** [comment]
Ok great, that all sounds good to me.  I'll get onto it in the morning.

**Comment by rawlingsj** [comment]
@ImJasonH @bobcatfish @abayer I was tempted to submit a few PR options for this but figured I'd start simple.  I've force pushed to this existing PR but to explain I've rebased which includes the tekton rename but doesn't include the https://github.com/knative/pkg/pull/235 package as suggested in an earlier comment.  My thinking is to keep the current `config/*.yaml` as is but I added docs to explain how to install into a custom namespace.  

If you want me to vendor the `knative/pkg` repo instead I can do that and update the controller and webhook deployments to use the downward api and update the docs to include that too.

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -118,6 +118,22 @@ kubectl create clusterrolebinding cluster-admin-binding \
   --user="${USER}"
 ```
 
+### Install in custom namespace
+1. To install into a different namespace you will need to modify resources in the `./config`
+    - remove all `namespace: tekton` resources
```
Comment (shashwathi) [inline DEVELOPMENT.md]: nit: Can you update this to  "remove all `namespace: tekton` references"

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -118,6 +118,22 @@ kubectl create clusterrolebinding cluster-admin-binding \
...
+    - delete the `namespace.yaml
```
Comment (shashwathi) [inline DEVELOPMENT.md]: Missing "`" at end of file name. Also can you link the config namespace file here?

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -118,6 +118,22 @@ kubectl create clusterrolebinding cluster-admin-binding \
...
+    - modify the `subjects.namespace` value to the desired namespace
```
Comment (shashwathi) [inline DEVELOPMENT.md]: Again can you link `clusterrolebinding` yaml here? I think thats what you are referring here

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -118,6 +118,22 @@ kubectl create clusterrolebinding cluster-admin-binding \
...
+    - add `downwardapi` entry to webhook and controller `deployment` resources and set current namespace.
```
Comment (shashwathi) [inline DEVELOPMENT.md]: Can you provide an example of how to add `downwardapi ` entry to both deployment resources?

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -118,6 +118,22 @@ kubectl create clusterrolebinding cluster-admin-binding \
...
+     
+     e.g.
+```yaml
```
Comment (shashwathi) [inline DEVELOPMENT.md]: I think you can remove the additional space in yaml example. Markdown render looks bit weird

File: DEVELOPMENT.md (hunk 1 above)
Comment (rawlingsj) [reply DEVELOPMENT.md]: done

File: DEVELOPMENT.md (hunk 2 above)
Comment (rawlingsj) [reply DEVELOPMENT.md]: done

File: DEVELOPMENT.md (hunk 3 above)
Comment (rawlingsj) [reply DEVELOPMENT.md]: done

File: DEVELOPMENT.md (hunk 4 above)
Comment (rawlingsj) [reply DEVELOPMENT.md]: done

File: DEVELOPMENT.md (hunk 5 above)
Comment (rawlingsj) [reply DEVELOPMENT.md]: done

**Comment by rawlingsj** [comment]
@shashwathi thanks for the comments.  Think I just resolved them, hopefully it now reads better.

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -118,6 +118,20 @@ kubectl create clusterrolebinding cluster-admin-binding \
   --user="${USER}"
 ```
 
+### Install in custom namespace
```
Comment (bobcatfish) [inline DEVELOPMENT.md]: i have one completely minor piece of feedback: usually we have an extra newline after a header.

i'm also happy to `lgtm` and let @mattmoor-sockpuppet fix that tho XD

Comment (rawlingsj) [reply DEVELOPMENT.md]: cool - thanks, I just added the extra newline anyways.

**Comment by abayer** [comment]
oh wait I don't have approve rights. heh.

**Pull request merged**